File: .claude/hooks/stop-loop.py
#!/usr/bin/env python3
"""
Autobot Stop Hook - Controls the autonomous loop continuation.

This is the heart of Autobot. It intercepts Claude's exit attempts and
determines whether to continue working based on task status and quality gates.

Hook Output Protocol:
  - Print JSON with "decision": "block" to prevent stop and re-inject prompt
  - Print JSON without "decision" or exit code 0 to allow stop
  - Exit code 2 for errors that should be shown to Claude
"""
import json
import sys
import os
from pathlib import Path
from datetime import datetime


def load_json(path: Path) -> dict | None:
    """Load JSON file if it exists."""
    if path.exists():
        try:
            return json.loads(path.read_text(encoding='utf-8'))
        except (json.JSONDecodeError, IOError):
            return None
    return None


def save_json(path: Path, data: dict) -> None:
    """Save data as JSON."""
    path.write_text(json.dumps(data, indent=2), encoding='utf-8')


def get_project_dir() -> Path:
    """Get the project directory from environment or cwd."""
    return Path(os.environ.get('CLAUDE_PROJECT_DIR', os.getcwd()))


def main():
    # Read hook input from stdin
    try:
        input_data = json.load(sys.stdin)
    except (json.JSONDecodeError, IOError):
        input_data = {}

    project_dir = get_project_dir()
    autobot_dir = project_dir / '.autobot'

    # Ensure autobot directory exists
    if not autobot_dir.exists():
        # No autobot setup - allow stop
        sys.exit(0)

    # Load state files
    task = load_json(autobot_dir / 'task.json')
    plan = load_json(autobot_dir / 'plan.json')
    metrics = load_json(autobot_dir / 'metrics.json') or {
        'current_iteration': 0,
        'max_iterations': 50,
        'consecutive_test_failures': 0
    }

    # Check if there's an active task
    if not task or task.get('status') == 'idle' or not task.get('id'):
        # No active task - allow stop
        sys.exit(0)

    # Check for pause flag
    if (autobot_dir / '.paused').exists():
        print(json.dumps({
            "continue": False,
            "stopReason": "Task paused. Use /resume to continue."
        }))
        sys.exit(0)

    # Check max iterations (safety limit)
    current_iteration = metrics.get('current_iteration', 0)
    max_iterations = metrics.get('max_iterations', 50)

    if current_iteration >= max_iterations:
        print(json.dumps({
            "continue": False,
            "stopReason": f"Safety limit reached: {max_iterations} iterations. Review progress and use /resume to continue."
        }))
        sys.exit(0)

    # Check for blocking failures (3+ consecutive test failures)
    consecutive_failures = metrics.get('consecutive_test_failures', 0)

    if consecutive_failures >= 3:
        print(json.dumps({
            "decision": "block",
            "reason": f"""PAUSE: {consecutive_failures} consecutive test failures detected.

Please review:
1. .autobot/observations.jsonl for failure patterns
2. .autobot/progress.md for any related learnings
3. The test output to understand the root cause

Options:
- Fix the issue and tests will auto-run
- Use /abort to stop and preserve state
- Use /pause to pause and resume later

What would you like to do?"""
        }))
        sys.exit(0)

    # Check subtask completion
    subtasks = plan.get('subtasks', []) if plan else []
    incomplete = [t for t in subtasks if t.get('status') != 'completed']
    completed = [t for t in subtasks if t.get('status') == 'completed']

    # Check for EXIT_SIGNAL in progress
    progress_path = autobot_dir / 'progress.md'
    progress_content = progress_path.read_text(encoding='utf-8') if progress_path.exists() else ""

    if "EXIT_SIGNAL: COMPLETE" in progress_content and not incomplete:
        # Task fully complete
        task['status'] = 'completed'
        save_json(autobot_dir / 'task.json', task)

        print(json.dumps({
            "continue": False,
            "stopReason": f"Task completed successfully! {len(completed)} subtasks finished."
        }))
        sys.exit(0)

    # If there are incomplete subtasks, continue
    if incomplete:
        next_task = incomplete[0]

        # Increment iteration counter
        metrics['current_iteration'] = current_iteration + 1
        metrics['last_activity'] = datetime.now().isoformat()
        save_json(autobot_dir / 'metrics.json', metrics)

        # Build continuation prompt
        prompt = f"""Continue working on the current task.

PROGRESS: {len(completed)}/{len(subtasks)} subtasks completed
ITERATION: {current_iteration + 1}/{max_iterations}

CURRENT SUBTASK: {next_task.get('title')}
DESCRIPTION: {next_task.get('description', 'No description')}
ACCEPTANCE CRITERIA:
{chr(10).join('- ' + c for c in next_task.get('acceptance_criteria', ['Complete the subtask']))}

INSTRUCTIONS:
1. Review .autobot/progress.md for any learned patterns
2. Write tests FIRST if this involves new functionality
3. Implement the change
4. Tests will auto-run after edits
5. If tests pass, update .autobot/plan.json to mark subtask complete
6. If all subtasks done, write "EXIT_SIGNAL: COMPLETE" to .autobot/progress.md

Quality over speed. Small, verifiable steps."""

        print(json.dumps({
            "decision": "block",
            "reason": prompt
        }))
        sys.exit(0)

    # No more subtasks but no EXIT_SIGNAL - prompt to complete
    print(json.dumps({
        "decision": "block",
        "reason": """All subtasks appear complete but EXIT_SIGNAL not found.

Please:
1. Verify all acceptance criteria are met
2. Run the full test suite one more time
3. If everything passes, append "EXIT_SIGNAL: COMPLETE" to .autobot/progress.md

This ensures proper task completion tracking."""
    }))


if __name__ == "__main__":
    main()

File: templates/claude/hooks/stop-loop.js
#!/usr/bin/env node
/**
 * Autobot Stop Hook - Controls the autonomous loop continuation.
 *
 * This is the heart of Autobot. It intercepts Claude's exit attempts and
 * determines whether to continue working based on task status and quality gates.
 *
 * Hook Output Protocol:
 *   - Print JSON with "decision": "block" to prevent stop and re-inject prompt
 *   - Print JSON without "decision" or exit code 0 to allow stop
 */
import fs from 'fs';
import path from 'path';

function loadJson(filePath) {
  try {
    return JSON.parse(fs.readFileSync(filePath, 'utf-8'));
  } catch {
    return null;
  }
}

function saveJson(filePath, data) {
  fs.writeFileSync(filePath, JSON.stringify(data, null, 2), 'utf-8');
}

function getProjectDir() {
  return process.env.CLAUDE_PROJECT_DIR || process.cwd();
}

function main() {
  // Drain stdin
  try {
    fs.readFileSync(0, 'utf-8');
  } catch {
    // No input
  }

  const projectDir = getProjectDir();
  const autobotDir = path.join(projectDir, '.autobot');

  if (!fs.existsSync(autobotDir)) {
    process.exit(0);
  }

  // Load state files
  const task = loadJson(path.join(autobotDir, 'task.json'));
  const plan = loadJson(path.join(autobotDir, 'plan.json'));
  const metrics = loadJson(path.join(autobotDir, 'metrics.json')) || {
    current_iteration: 0,
    max_iterations: 50,
    consecutive_test_failures: 0,
  };

  // Check if there's an active task
  if (!task || task.status === 'idle' || !task.id) {
    process.exit(0);
  }

  // Check for pause flag
  if (fs.existsSync(path.join(autobotDir, '.paused'))) {
    console.log(JSON.stringify({
      continue: false,
      stopReason: 'Task paused. Use /resume to continue.',
    }));
    process.exit(0);
  }

  // Check max iterations
  const currentIteration = metrics.current_iteration || 0;
  const maxIterations = metrics.max_iterations || 50;

  if (currentIteration >= maxIterations) {
    console.log(JSON.stringify({
      continue: false,
      stopReason: `Safety limit reached: ${maxIterations} iterations. Review progress and use /resume to continue.`,
    }));
    process.exit(0);
  }

  // Check for blocking failures
  const consecutiveFailures = metrics.consecutive_test_failures || 0;

  if (consecutiveFailures >= 3) {
    console.log(JSON.stringify({
      decision: 'block',
      reason: `PAUSE: ${consecutiveFailures} consecutive test failures detected.\n\nPlease review:\n1. .autobot/observations.jsonl for failure patterns\n2. .autobot/progress.md for any related learnings\n3. The test output to understand the root cause\n\nOptions:\n- Fix the issue and tests will auto-run\n- Use /abort to stop and preserve state\n- Use /pause to pause and resume later\n\nWhat would you like to do?`,
    }));
    process.exit(0);
  }

  // Check subtask completion
  const subtasks = (plan && plan.subtasks) ? plan.subtasks : [];
  const incomplete = subtasks.filter(t => t.status !== 'completed');
  const completed = subtasks.filter(t => t.status === 'completed');

  // Check for EXIT_SIGNAL in progress
  const progressPath = path.join(autobotDir, 'progress.md');
  let progressContent = '';
  if (fs.existsSync(progressPath)) {
    progressContent = fs.readFileSync(progressPath, 'utf-8');
  }

  if (progressContent.includes('EXIT_SIGNAL: COMPLETE') && incomplete.length === 0) {
    task.status = 'completed';
    saveJson(path.join(autobotDir, 'task.json'), task);

    console.log(JSON.stringify({
      continue: false,
      stopReason: `Task completed successfully! ${completed.length} subtasks finished.`,
    }));
    process.exit(0);
  }

  // If there are incomplete subtasks, continue
  if (incomplete.length > 0) {
    const nextTask = incomplete[0];

    // Increment iteration counter
    metrics.current_iteration = currentIteration + 1;
    metrics.last_activity = new Date().toISOString();
    saveJson(path.join(autobotDir, 'metrics.json'), metrics);

    const criteria = (nextTask.acceptance_criteria || ['Complete the subtask'])
      .map(c => `- ${c}`)
      .join('\n');

    const prompt = `Continue working on the current task.\n\nPROGRESS: ${completed.length}/${subtasks.length} subtasks completed\nITERATION: ${currentIteration + 1}/${maxIterations}\n\nCURRENT SUBTASK: ${nextTask.title}\nDESCRIPTION: ${nextTask.description || 'No description'}\nACCEPTANCE CRITERIA:\n${criteria}\n\nINSTRUCTIONS:\n1. Review .autobot/progress.md for any learned patterns\n2. Write tests FIRST if this involves new functionality\n3. Implement the change\n4. Tests will auto-run after edits\n5. If tests pass, update .autobot/plan.json to mark subtask complete\n6. If all subtasks done, write "EXIT_SIGNAL: COMPLETE" to .autobot/progress.md\n\nQuality over speed. Small, verifiable steps.`;

    console.log(JSON.stringify({
      decision: 'block',
      reason: prompt,
    }));
    process.exit(0);
  }

  // No more subtasks but no EXIT_SIGNAL
  console.log(JSON.stringify({
    decision: 'block',
    reason: 'All subtasks appear complete but EXIT_SIGNAL not found.\n\nPlease:\n1. Verify all acceptance criteria are met\n2. Run the full test suite one more time\n3. If everything passes, append "EXIT_SIGNAL: COMPLETE" to .autobot/progress.md\n\nThis ensures proper task completion tracking.',
  }));
}

main();
